### PR TITLE
Possibly faster RNNs

### PR DESCRIPTION
**Pull request by fchollet**
- make it possible to use RNN dropout with TensorFlow without specifying a batch size
- move initial RNN matrix multiplications outside of recurrent loop / scan -> runs faster

@yaringal @sunil-at-gh any feedback?

### CODE REVIEW

**Comment by farizrahman4u** [comment]
`x = K.reshape(x, (-1, timesteps, output_dim))`

What if timesteps is `None`?

**Comment by farizrahman4u** [comment]
:+1:

**Comment by sunil-at-gh** [comment]
It looks like my changes do not work with TensorFlow, but your might.
I had two main motivations for the changes I made, prompted by my porting some custom Theano code to Keras and finding the Keras version taking more than twice the time:
- Basic training time improvements. Performing pre- and post- recurrence computations outside the Scan were the biggest components, although other were smaller but significant
- Making the initial state optionally trainable.

I did find the time benchmark script useful … it also checked that I got (almost) the same loss as the non-streamlined versions. I have it up as a Gist at:
https://gist.github.com/sunil-at-gh/0513c1d30cb378c66b0d <https://gist.github.com/sunil-at-gh/0513c1d30cb378c66b0d> 

_
Sunil

**Comment by sunil-at-gh** [comment]
Tested my 'benchmark' script on your changes. It compares the previous implementation against the new implementation of each RNN layer type, with and without dropout. Without dropout things look good. With dropout, I see different loss numbers. Here is an excerpt from the output (Theano, cpu):

```
Running Training-time test for RNN with dropout = 0 ...
nbr time steps = 20, width of input to RNN at each step = 1000
Building models ...
Training model_k for 10 epochs ...
Training model_s for 10 epochs ...
Loss for model_k = 8.90779015754e-05
Loss for model_s = 8.90779032488e-05
Run times for 10 epochs, dropout = 0:
  model with original RNN:    32.2 sec
  model with Streamlined RNN: 18.8 sec
  run-time improvement = 41.5%

Running Training-time test for RNN with dropout = 0.7 ...
nbr time steps = 20, width of input to RNN at each step = 1000
Building models ...
Training model_k for 10 epochs ...
Training model_s for 10 epochs ...
Loss for model_k = 0.000429985868977
Loss for model_s = 0.00034108853084
Run times for 10 epochs, dropout = 0.7:
  model with original RNN:    39.6 sec
  model with Streamlined RNN: 29.2 sec
  run-time improvement = 26.2%
```

Notice the loss numbers in the second section are slightly different.

**Comment by NasenSpray** [comment]
Please consider making this optional as there are actually cases in which this leads to significant slow-downs (Theano, GPU). I can't remember right now where I encountered this, but I should have an example somewhere and am going to try find it.

**Comment by farizrahman4u** [comment]
How can getting stuff out of scan make things slow?

**Comment by yaringal** [comment]
Looks good to me. I noticed this though:
Lines 322, 485: no return value

**Comment by fchollet** [comment]
Thanks guys. Will merge once the tests are complete.

> How can getting stuff out of scan make things slow?

I don't think that's possible. Unless... the bigger matrix multiplication ends up being routed on CPU rather than on GPU.

**Comment by sunil-at-gh** [comment]
I added a predictions comparison check to my test script, and the two model versions (before and after changes) return different predictions. So the change is not just in training.

I noticed that yaringal's version had a mask of 1's and 0's during training, and multiplied the input with (1 - dropout_pr) at prediction time. The version on this branch takes the other approach, where the masks at training time have 1/(1 - dropout_pr) instead of a 1.

I did a quick edit on SimpleRNN and modified it to produce masks in the same style as yaringal's version. I still do not get the same loss or predictions, but they are now much closer. Perhaps the random number generators are being called in a different order, and so produce slightly different numbers?

**Comment by fchollet** [comment]
> the two model versions (before and after changes) return different predictions

The dropout makes the model non-deterministic, so I would not expect the same predictions. Since the source of the randomness has changed, even seeding will give different results.

> The version on this branch takes the other approach, where the masks at training time have 1/(1 - dropout_pr) instead of a 1.

Correct. This is the way dropout is handled everywhere else in Keras. It is advantageous because it means that the test time behavior of a trained network does not change if your remove dropout from the network.

**Comment by sunil-at-gh** [comment]
These changes look good. I will start using them.
Learnt some more about the differences between Theano and TF :-)

Yes, I too prefer the dropout approach used in Keras. Easier to compare loss values and lr with and without dropout this way.

**Comment by sunil-at-gh** [comment]
If this change now supports unspecified batch size in TensorFlow, did you also want to update test_recurrent.py to use input_shape instead of batch_input_shape in the dropout tests?

**Comment by lemuriandezapada** [comment]
Funny enough, I'm not seeing any speed improvements on my LSTMs for my Titan X with this change, but I am seeing at least a doubling of memory usage.

**Comment by NasenSpray** [comment]
> > How can getting stuff out of scan make things slow?

> I don't think that's possible. Unless... the bigger matrix multiplication ends up being routed on CPU rather than on GPU.

```
(X_train, y_train), (X_test, y_test) = mnist.load_data()
Y_train = np_utils.to_categorical(y_train, 10)
model = Sequential()
model.add(LSTM(100, input_shape=(784, 1), return_sequences=True))
model.add(LSTM(100, return_sequences=False))
model.add(Dense(10, activation='softmax'))
model.compile(Adam(), 'categorical_crossentropy')
model.fit(X_train.reshape(-1, 784, 1), Y_train.reshape(-1, 10), batch_size=64, nb_epoch=1)
```

GTX 970, Theano backend:
- pre merge: 1600s/epoch, ~800MB VRAM peak
- post merge: 3200s/epoch, ~1150MB VRAM peak

**Comment by fchollet** [comment]
People having performance issues: try adjusting the batch size. There is no doubt that doing as a single matrix multiplication what used to be `timesteps` separate matrix multiplications will yield better performance, unless you are in some memory-management related edge-case.

**Comment by NasenSpray** [comment]
@fchollet:
> There is no doubt that doing as a single matrix multiplication what used to be timesteps separate matrix multiplications will yield better performance, unless you are in some memory-management related edge-case.

...did you actually meassure the performance? CPU runs _don't_ count.
```
X_train = np.random.uniform(size=(20000, 20, 500))
Y_train = np.random.uniform(size=(20000, 10))
model = Sequential()
model.add(LSTM(100, input_shape=(20, 500), return_sequences=True))
model.add(LSTM(100, return_sequences=False))
model.add(Dense(10, activation='softmax'))
model.compile(Adam(), 'categorical_crossentropy')
model.fit(X_train, Y_train.reshape(-1, 10), batch_size=256, nb_epoch=1)
```
- pre merge: 6s
- post merge: 10s

It takes longer and needs more memory, great improvement!

**Comment by lemuriandezapada** [comment]
Any update on this? The performance drop is quite a hit.

**Comment by lemuriandezapada** [comment]
@fchollet 
"People having performance issues: try adjusting the batch size. There is no doubt that doing as a single matrix multiplication what used to be timesteps separate matrix multiplications will yield better performance, unless you are in some memory-management related edge-case."

^ That's not necessarily the case. CUDA performance is pretty much measured in how much math you can compute at once in one matrix multiplication. If your batch sizes are big enough, then it doesn't matter that you do it iteratively over the time dimension. But the large memory footprint will force you to decrease your batch sizes and thus kill efficiency.

**Comment by bayesrule** [comment]
@fchollet 
**Really** hope you can take this seriously. I tried both the latest 0.3.3 keras (since keras-1 is not yet completed) and the pre-PR version. Running the same **imdb_bidirectional_lstm.py example**,

it turns out that the performance does decrease quite a lot on my GTX 970M gpu with CUDA 7.5 and Theano 0.8.1.

0.3.3         --> 299s per epoch
pre-PR     --> 116s per epoch

Many thanks!
